### DISCUSSION TITLE
How to build a basic firmware for nRF52840 ?

### DISCUSSION THREAD

**Discussion by bbadrignans** [ffenix113/zigbee_home#37] [Q&A] [answered]
I'm interested to try this project, but I can't figure out how to build something to flash on my nRF52840 Dongle.
I'm new to Go and Zephyr, so sorry if I ask dumb questions. Maybe I'm not the only one, so I can try to enhance the documentation if you help me.

Information : I use Debian 12. I'm on the dev branch.  I installed go1.21.4 from https://go.dev/dl/.

If I try the command from documentation at the root directory : 
```
go run ./cli/cmd/zigbee --config ./zigbee_test.yml firmware --workdir ./firmware build
go: cannot find main module, but found .git/config in [...]/zigbee_home
        to create a module there, run:
        go mod init
```
I tried to `go mod init` but I suppose it is not what to do.
Then I succeed to build zigbee command line : 
```
cd zigbee_home/cli/cmd/zigbee
go build
```
I can run it, after commenting lines 107 to 110 in zigbee.yml (leds pins), then installing west with pip
```
cd zigbee_home/cli/
python -m venv venv
source venv/bin/activate
pip3 install -U west
go run ./cmd/zigbee --config ./zigbee.yml firmware --workdir ./cmd/zigbee/firmware build
2024/03/24 21:48:13 generator.go:106: Device: "nrf52840dongle_nrf52840", selected bootloader: "nrf52_legacy", forced bootloader: false
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
west: unknown command "build"; do you need to run this inside a workspace?
2024/03/24 21:48:13 main.go:30: build firmware: run command "west": exit status 1
exit status 1
```
I then try to run west manually, but it failed : 
```
cd zigbee_home/cli/cmd/zigbee/firmware
west init
west build --pristine --board nrf52840dongle_nrf52840 --no-sysbuild --build-dir build -- -DNCS_TOOLCHAIN_VERSION=NONE -DDTC_OVERLAY_FILE=app.overlay -DCONF_FILE=prj.conf
[...]
-- Zephyr version: 3.6.99 ([...]/zigbee_home/cli/cmd/zigbee/firmware/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
CMake Warning at zephyr/cmake/modules/boards.cmake:105 (message):
  Deprecated BOARD=nrf52840dongle_nrf52840 specified, board automatically
  changed to: nrf52840dongle.
Call Stack (most recent call first):
  zephyr/cmake/modules/zephyr_default.cmake:132 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:18 (find_package)

-- Board: nrf52840dongle, qualifiers: nrf52840
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
CMake Error at zephyr/cmake/modules/FindZephyr-sdk.cmake:109 (find_package):
  Could not find a package configuration file provided by "Zephyr-sdk"
  (requested version 0.16) with any of the following names:

    Zephyr-sdkConfig.cmake
    zephyr-sdk-config.cmake

  Add the installation prefix of "Zephyr-sdk" to CMAKE_PREFIX_PATH or set
  "Zephyr-sdk_DIR" to a directory containing one of the above files.  If
  "Zephyr-sdk" provides a separate development package or SDK, be sure it has
  been installed.
```
I think I missed something. 
Can someone help me to find what I do wrong ?

**Comment by ffenix113** [comment, marked as answer]
Hello @bbadrignans !

Thank you for the interest in the project and the question!
I agree that the documentation for the project is lacking, and it is something that will be improved soon(I am working on ADC sensor for now).

Installation process can be something like this:
1) Install nRF Connect SDK. The step is explained [here](https://ffenix113.github.io/zigbee_home/using_the_cli/building_the_firmware/#preparing-environment).
2) Install the binary. As you mentioned that you have pulled the repo - please follow [this](https://ffenix113.github.io/zigbee_home/install_zigbee_home_cli/#running-from-source-without-installation) documentation.
3) Prepare the configuration file. You probably also have that, as you said.
4) Build & flash firmware. The docs are less exact on this topic, but you can follow this links:
4.1) To build the firmware - https://ffenix113.github.io/zigbee_home/using_the_cli/#build-the-firmware. Please know that CLI can now build the firmware without going to VS Code and starting the terminal from there, so you can skip the note written on the page.
4.2) To flash the firmware - https://ffenix113.github.io/zigbee_home/using_the_cli/#flashing-the-firmware.

While this is also no replacement for a good docs page - I hope that you find this useful. Also please check the docs from the links.

If you have further issues or questions - please post them here.

**Reply by bbadrignans** [reply]
I succeed to build and flash a firmware, thank you for your help.
To be honest, I didn't see the "docs" folder in the git repo...

I used the latest nRF Connect SDK (v2.6.0) and toolchain.
I needed to change some code in cli to achieve the build. I don't know if it is due to my environment.
I will propose a patch that you may review and check in your environment.

![image](https://github.com/ffenix113/zigbee_home/assets/534410/067a4c51-89b3-4f9c-b27d-719dd6b34eb6)

**Reply by ffenix113** [reply]
Nice! Glad to hear that I was of some help!

I also need to add configuration options for device information(i.e. manufacturer, device id, etc), as it also will be a nice differentiation for generated devices.

About MR: I will take a look at it ASAP. I do believe that env configuration can be "sub-optimal" for sure. Would need to test on v2.5.X and v2.6.X versions though.

**Comment by bbadrignans** [comment]
Thanks a lot, I will try that, keep you informed and if I can, propose documentation improvements.
I don't want to disturb you since I'm interested in ADC sensor 😄

**Comment by Hedda** [comment]
Suggest to also use project's wiki for documentation if not revision/version dependent -> https://github.com/ffenix113/zigbee_home/wiki

**Reply by ffenix113** [reply]
It would be nice to do so. Technically, we can probably even try to share the documentation in Wiki and on the site using this: https://docs.github.com/en/communities/documenting-your-project-with-wikis/adding-or-editing-wiki-pages#adding-or-editing-wiki-pages-locally

The biggest problem, for now, is that I think only I can add the docs, as the project does not have actual users yet that could help by updating the wiki with their project knowledge.
